### PR TITLE
api: Disable body limit middleware for admin endpoints.

### PR DESCRIPTION
**Pull request by winder**
## Summary

Disable body limit middleware for admin endpoints.

## Test Plan

Unit test.

### CODE REVIEW

**Review by jannotti — COMMENTED** [review]
I'm not sure exactly what the behavior of this middleware is, but I think I'd like to see _some_ limit, even on admin access. Otherwise, I suspect algod can be crashed by a request that just never ends. Eventually the server exhausts memory by reading arbitrary data.

I guess you could say, "well, admins can crash the server" and I'm actually somewhat sympathetic to that if it's annoying to implement. But if it's just a different limit in the admin middleware, let's do it, with maybe 100mb?

**Comment by winder** [comment]
> I'm not sure exactly what the behavior of this middleware is, but I think I'd like to see _some_ limit, even on admin access. Otherwise, I suspect algod can be crashed by a request that just never ends. Eventually the server exhausts memory by reading arbitrary data.
> 
> I guess you could say, "well, admins can crash the server" and I'm actually somewhat sympathetic to that if it's annoying to implement. But if it's just a different limit in the admin middleware, let's do it, with maybe 100mb?

"admins can crash the server" was my thinking here. It would also be simple to add different middlewares for each set of endpoints, but is 100MB is the right number? Without some rational about why there should be a limit and what it should be, I'd rather leave it open.

**Comment by jannotti** [comment]
10mb had no rationale beyond "not large enough to exhaust memory, and should be enough for anything"

But we forgot part keys. So let's pick a value that _still_ won't exhaust memory, but is big enough for the number of part keys we want to support.

By the way, what I'm _really_ worried about is that leaving it open means that a request could be arbitrarily large, even before it is authenticated as admin.

**Comment by winder** [comment]
> By the way, what I'm _really_ worried about is that leaving it open means that a request could be arbitrarily large, even before it is authenticated as admin.

I don't know if this alleviates the concern, but the token check happens ahead of any reading. So you don't need to worry about this part.

Since it's rate limiting is implemented as a middleware, the enforcement needs to happen inline to the handler call chain. It does so using two mechanisms, each of these happen ahead of our handler code:
1) check the content size and return if it reports the size is too large (a malicious request would not set the content size correctly, so this doesn't help with your concern)
2) the response body reader is wrapped with a rate limited reader that returns an error once the configured number of bytes is reached.

**Review by shiqizng — DISMISSED** [review]

**Review by Eric-Warehime — DISMISSED** [review]
LGTM but needs rebase

```
indexer_test.go:88: 
        	Error Trace:	/opt/cibuild/project/node/indexer/indexer_test.go:88
        	            				/opt/cibuild/project/node/indexer/suite.go:132
        	            				/opt/cibuild/project/node/indexer/indexer_test.go:164
```

File: daemon/algod/api/server/v2/test/handlers_test.go [hunk 1]
```diff
@@ -2203,3 +2204,37 @@ func TestDeltasForTxnGroup(t *testing.T) {
...
+	dummyShutdownChan := make(chan struct{})
+	l, err := net.Listen("tcp", ":0") // create listener so requests are buffered
+	e := server.NewRouter(logging.TestingLog(t), mockNode, dummyShutdownChan, "", "", l, 1000)
+	go e.Start(":0")
+	defer e.Close()
+
+	// Admin API call greater than max body bytes should succeed
+	assert.Equal(t, "10MB", server.MaxRequestBodyBytes)
+	stringReader := strings.NewReader(strings.Repeat("a", 50_000_000))
+	fmt.Println(e.Listener)
```
Comment (algochoi) [inline daemon/algod/api/server/v2/test/handlers_test.go:2223]: Do we print here on purpose?

**Review by algochoi — DISMISSED** [review]
Nice, LGTM

File: daemon/algod/api/server/v2/test/handlers_test.go (hunk 1 above)
Comment (winder) [reply daemon/algod/api/server/v2/test/handlers_test.go]: No, thanks.

**Review by algochoi — APPROVED** [review]
👍

**Review by Eric-Warehime — APPROVED** [review]

**Comment by onetechnical** [comment]
> "admins can crash the server" was my thinking here. It would also be simple to add different middlewares for each set of endpoints, but is 100MB is the right number? Without some rational about why there should be a limit and what it should be, I'd rather leave it open.

10 million rounds (at the long end) is what we would recommend as maximum if you do NOT have a large amount of stake:

4.3s era 10m rounds = 497.7 days
3.7s era 10m rounds = 428.2 days
3.2s era 10m rounds = 370.4 days

I'm not inclined to recommend increasing this number, as longer than 1 year still seems too long.

That said, a 10m round participation key with dilution of 3162 is 192 MB, so I think an admin cap of, say, 250 MB might be sufficient. However, as this is limited to admin operations, I'm also inclined to leave it out, but your mileage may vary.

**Pull request merged**